tidy3d/plugins/adjoint/components/geometry.py

```diff
@@ -0,0 +1,279 @@
+                    continue
+
+                # get the dimensions and edge values on the plane of this surface
+                _, dims_plane = self.pop_axis("xyz", axis=dim_index)
+                _, mins_plane = self.pop_axis(rmin, axis=dim_index)
+                _, maxs_plane = self.pop_axis(rmax, axis=dim_index)
+
+                # construct differential area value and coordinates evenly spaced along this surface
+                d_area = 1.0
+                area_coords = {}
+                for dim_plane, min_edge, max_edge in zip(dims_plane, mins_plane, maxs_plane):
+
+                    # if there is no thickness along this dimension, skip it
+                    length_edge = max_edge - min_edge
+                    if length_edge == 0:
+                        continue
+
+                    # update the differential area value
+                    d_len = length_edge / NUM_CELLS_DIM
+                    d_area *= d_len
+
+                    # construct evenly spaced coordinates along this dimension
+                    interp_vals = np.linspace(
+                        min_edge + d_len / 2, max_edge - d_len / 2, NUM_CELLS_DIM
+                    )
+                    area_coords[dim_plane] = interp_vals
+
+                # for each field component
+                for field_cmp_dim in "xyz":
+
+                    # select the forward and adjoint E fields along this edge
+                    field_name = "E" + field_cmp_dim
+                    e_fwd = grad_data_fwd.field_components[field_name].isel(f=0)
+                    e_adj = grad_data_adj.field_components[field_name].isel(f=0)
+
+                    # select the permittivity data
+                    eps_field_name = f"eps_{field_cmp_dim}{field_cmp_dim}"
+                    eps_data = grad_data_eps.field_components[eps_field_name].isel(f=0)
+
+                    # get the permittivity values just inside and outside the edge
+                    n_cells_in = 2
+                    isel_out = 0 if min_max_index == 0 else -1
+                    isel_ins = n_cells_in if min_max_index == 0 else -n_cells_in - 1
+                    eps2 = eps_data.isel(**{dim_normal: isel_out})
+                    eps1 = eps_data.isel(**{dim_normal: isel_ins})
+
+                    # get gradient contribution for normal component using D field
+                    if field_cmp_dim == dim_normal:
+
+                        # construct normal D fields, dot together at surface
+                        d_fwd = e_fwd * eps_data
+                        d_adj = e_adj * eps_data
+                        d_normal = (d_fwd * d_adj).interp(**normal_coord)
+
+                        # compute adjoint contribution using perturbation theory for shifting bounds
+                        delta_eps_inv = 1.0 / eps1 - 1.0 / eps2
+                        d_integrand = -(delta_eps_inv * d_normal).interp(**area_coords).real
```

Comment:
Could you explain a bit on why it's taking the real part? Is it assuming real permittivity values?

Reply:
It's simply from the fact that the final objective function is real.